ClassNotFoundException: byte[] for serialized valued.

Hi. I have some problem with play-redis and I think this is a bug.
This Plugin is work fine if we didn't use custom serialization. 
This is our scenario. The value serialized before set and deserialized after get. 
The value after serialization is byte[] and the ClassTag stored in Redis is byte[]. 
When we called get we got this exception:

```
Caused by: java.lang.ClassNotFoundException: byte[]
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:466)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:566)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:499)
	at play.api.cache.redis.impl.JavaRedis.decodeClassTag$1(JavaRedis.scala:60)
	at play.api.cache.redis.impl.JavaRedis.$anonfun$getOrElse$1(JavaRedis.scala:61)
	at scala.Option.map(Option.scala:146)
	at play.api.cache.redis.impl.JavaRedis.decodedClassTag$1(JavaRedis.scala:61)
	at play.api.cache.redis.impl.JavaRedis.$anonfun$getOrElse$2(JavaRedis.scala:63)
	at scala.util.Success.$anonfun$map$1(Try.scala:251)
	at scala.util.Success.map(Try.scala:209) 
```
If the byte[] is incorrect type for value, this method should be fix : 
```

  def set( key: String, value: scala.Any, duration: Duration ): Future[ Done ] = {
    import dsl._

    Future.sequence(
      Seq(
        // set the value
        internal.set( key, value, duration ),
        // and set its type to be able to read it
        internal.set( s"classTag::$key", if ( value == null ) "" else value.getClass.getCanonicalName, duration )
      )
    ).map {
      case Seq( done, _ ) => done
    }
  }
```

Maybe if you define another set method that get class type our problem resolved! 

Version of play-redis = 2.1.2
Version of Scala = 2.12.6
Version of Play = 2.6.15

I write simple module in play that run this code : 
```
package services;

import play.Logger;
import play.cache.SyncCacheApi;

import javax.inject.Inject;
import javax.inject.Singleton;

/**
 * @author vahid
 */
@Singleton
public class KryoBindService {

    private final SyncCacheApi syncCacheApi;
    protected final Logger.ALogger logger = Logger.of(getClass());


    @Inject
    public KryoBindService(SyncCacheApi syncCacheApi) {
        this.syncCacheApi = syncCacheApi;
        logger.info("start get and set byte[] to play redis");

        try {


            syncCacheApi.set("string", "vahid");
            System.out.println("string:" + syncCacheApi.get("string"));

            syncCacheApi.set("bytes", new byte[0]);
            System.out.println("bytes:" + String.valueOf(syncCacheApi.get("bytes")));

            logger.info("finish get and set byte[] to play redis");
        } catch (Throwable e) {
            e.printStackTrace();
        }
    }
}

```

and got this error : 
```
[info] 2018-07-03 14:13:18,099 s.KryoBindService - start get and set byte[] to play redis
string:vahid
java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.ClassNotFoundException: byte[]
	at play.cache.DefaultSyncCacheApi.blocking(DefaultSyncCacheApi.java:92)
	at play.cache.DefaultSyncCacheApi.get(DefaultSyncCacheApi.java:35)
	at services.KryoBindService.<init>(KryoBindService.java:31)
	at services.KryoBindService$$FastClassByGuice$$be9a370b.newInstance(<generated>)
	at com.google.inject.internal.DefaultConstructionProxyFactory$FastClassProxy.newInstance(DefaultConstructionProxyFactory.java:89)
	at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:111)
	at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:90)
	at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:268)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1092)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
	at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:194)
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:41)
	at com.google.inject.internal.InternalInjectorCreator$1.call(InternalInjectorCreator.java:205)
	at com.google.inject.internal.InternalInjectorCreator$1.call(InternalInjectorCreator.java:199)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1085)
	at com.google.inject.internal.InternalInjectorCreator.loadEagerSingletons(InternalInjectorCreator.java:199)
	at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:180)
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:110)
	at com.google.inject.Guice.createInjector(Guice.java:99)
	at com.google.inject.Guice.createInjector(Guice.java:84)
	at play.api.inject.guice.GuiceBuilder.injector(GuiceInjectorBuilder.scala:185)
	at play.api.inject.guice.GuiceApplicationBuilder.build(GuiceApplicationBuilder.scala:137)
	at play.api.inject.guice.GuiceApplicationLoader.load(GuiceApplicationLoader.scala:21)
	at play.core.server.DevServerStart$$anon$1.$anonfun$reload$3(DevServerStart.scala:174)
	at play.utils.Threads$.withContextClassLoader(Threads.scala:21)
	at play.core.server.DevServerStart$$anon$1.reload(DevServerStart.scala:171)
	at play.core.server.DevServerStart$$anon$1.get(DevServerStart.scala:124)
	at play.core.server.AkkaHttpServer.handleRequest(AkkaHttpServer.scala:222)
	at play.core.server.AkkaHttpServer.$anonfun$createServerBinding$1(AkkaHttpServer.scala:137)
	at akka.stream.impl.fusing.MapAsync$$anon$25.onPush(Ops.scala:1194)
	at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:519)
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:411)
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:585)
	at akka.stream.impl.fusing.GraphInterpreterShell$AsyncInput.execute(ActorGraphInterpreter.scala:469)
	at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:560)
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:742)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:757)
	at akka.actor.Actor.aroundReceive(Actor.scala:517)
	at akka.actor.Actor.aroundReceive$(Actor.scala:515)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:667)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:590)
	at akka.actor.ActorCell.invoke(ActorCell.scala:559)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.util.concurrent.ExecutionException: java.lang.ClassNotFoundException: byte[]
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2022)
	at scala.concurrent.java8.FuturesConvertersImpl$CF.super$get(FutureConvertersImpl.scala:84)
	at scala.concurrent.java8.FuturesConvertersImpl$CF.$anonfun$get$2(FutureConvertersImpl.scala:84)
	at akka.dispatch.MonitorableThreadFactory$AkkaForkJoinWorkerThread$$anon$3.block(ThreadPoolBuilder.scala:167)
	at akka.dispatch.forkjoin.ForkJoinPool.managedBlock(ForkJoinPool.java:3641)
	at akka.dispatch.MonitorableThreadFactory$AkkaForkJoinWorkerThread.blockOn(ThreadPoolBuilder.scala:165)
	at scala.concurrent.package$.blocking(package.scala:142)
	at scala.concurrent.java8.FuturesConvertersImpl$CF.get(FutureConvertersImpl.scala:84)
	at play.cache.DefaultSyncCacheApi.blocking(DefaultSyncCacheApi.java:86)
	... 49 more
Caused by: java.lang.ClassNotFoundException: byte[]
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:466)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:566)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:499)
	at play.api.cache.redis.impl.JavaRedis.decodeClassTag$1(JavaRedis.scala:60)
	at play.api.cache.redis.impl.JavaRedis.$anonfun$getOrElse$1(JavaRedis.scala:61)
	at scala.Option.map(Option.scala:146)
	at play.api.cache.redis.impl.JavaRedis.decodedClassTag$1(JavaRedis.scala:61)
	at play.api.cache.redis.impl.JavaRedis.$anonfun$getOrElse$2(JavaRedis.scala:63)
	at scala.util.Success.$anonfun$map$1(Try.scala:251)
	at scala.util.Success.map(Try.scala:209)
	at scala.concurrent.Future.$anonfun$map$1(Future.scala:288)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:29)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:29)
	at scala.concurrent.impl.CallbackRunnable.run$$$capture(Promise.scala:60)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala)
	at play.core.j.HttpExecutionContext$$anon$2.run(HttpExecutionContext.scala:56)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)

```
The"classTag::bytes" for bytes key is byte[]. 
The error is exist.
I used incremental and default strategy and the problem exist for Kryo.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,10 @@
 
 ## Changelog
 
+### [:link: 2.2.0](https://github.com/KarelCemus/play-redis/tree/2.2.0)
+
+Support of plain arrays in JavaRedis [#176](https://github.com/KarelCemus/play-redis/pull/176).
+
 ### [:link: 2.1.2](https://github.com/KarelCemus/play-redis/tree/2.1.2)
 
 JDK 10 compatibility. Replace deprecated com.sun.misc.BASE64* usages with jdk8 java.util.Base64 [#170](https://github.com/KarelCemus/play-redis/pull/170).

--- a/src/main/scala/play/api/cache/redis/impl/JavaRedis.scala
+++ b/src/main/scala/play/api/cache/redis/impl/JavaRedis.scala
@@ -33,7 +33,7 @@ private[ impl ] class JavaRedis( internal: CacheAsyncApi, environment: Environme
         // set the value
         internal.set( key, value, duration ),
         // and set its type to be able to read it
-        internal.set( s"classTag::$key", if ( value == null ) "" else value.getClass.getCanonicalName, duration )
+        internal.set( s"classTag::$key", classTagOf( value ), duration )
       )
     ).map {
       case Seq( done, _ ) => done
@@ -57,8 +57,7 @@ private[ impl ] class JavaRedis( internal: CacheAsyncApi, environment: Environme
     implicit val context = HttpExecutionContext.fromThread( runtime.context )
     // get the tag and decode it
     def getClassTag = internal.get[ String ]( s"classTag::$key" )
-    def decodeClassTag( name: String ): ClassTag[ T ] = if ( name == "" ) ClassTag.Null.asInstanceOf[ ClassTag[ T ] ] else ClassTag( environment.classLoader.loadClass( name ) )
-    def decodedClassTag( tag: Option[ String ] ) = tag.map( decodeClassTag )
+    def decodedClassTag( tag: Option[ String ] ) = tag.map( classTagFrom[ T ] )
     // if tag is defined, get Option[ value ] otherwise None
     def getValue = getClassTag.map( decodedClassTag ).flatMap {
       case Some( ClassTag.Null ) => Future.successful( Some( null.asInstanceOf[ T ] ) )
@@ -81,6 +80,15 @@ private[ impl ] class JavaRedis( internal: CacheAsyncApi, environment: Environme
   }
 
   def removeAll( ) = internal.invalidate().toJava
+
+  protected def classTagOf( value: Any ): String = {
+    if ( value == null ) "" else value.getClass.getCanonicalName
+  }
+
+  protected def classTagFrom[ T ]( tag: String ): ClassTag[ T ] = {
+    if ( tag == "" ) ClassTag.Null.asInstanceOf[ ClassTag[ T ] ]
+    else ClassTag( classTagNameToClass( tag, environment ) )
+  }
 }
 
 private[ impl ] object JavaRedis {
@@ -93,4 +101,19 @@ private[ impl ] object JavaRedis {
   private[ impl ] implicit class ScalaCompatibility[ T ]( val future: CompletionStage[ T ] ) extends AnyVal {
     @inline def toScala: Future[ T ] = FutureConverters.toScala( future )
   }
+
+  // $COVERAGE-OFF$
+  /** java primitives are serialized into their type names instead of classes */
+  private[ impl ] def classTagNameToClass( name: String, environment: Environment ): Class[ _ ] = name match {
+    case "boolean[]" => classOf[ Array[ java.lang.Boolean ] ]
+    case "byte[]" => classOf[ Array[ java.lang.Byte ] ]
+    case "char[]" => classOf[ Array[ java.lang.Character ] ]
+    case "short[]" => classOf[ Array[ java.lang.Short ] ]
+    case "int[]" => classOf[ Array[ java.lang.Integer ] ]
+    case "long[]" => classOf[ Array[ java.lang.Long ] ]
+    case "float[]" => classOf[ Array[ java.lang.Float ] ]
+    case "double[]" => classOf[ Array[ java.lang.Double ] ]
+    case clazz => environment.classLoader.loadClass( clazz )
+  }
+  // $COVERAGE-ON$
 }

--- a/src/test/java/play/api/cache/redis/JavaTypes.java
+++ b/src/test/java/play/api/cache/redis/JavaTypes.java
@@ -1,0 +1,10 @@
+package play.api.cache.redis;
+
+/**
+ * @author Karel Cemus
+ */
+public class JavaTypes {
+
+    public static final byte byteValue = 5;
+    public static final byte[] bytesValue = new byte[]{1, 2, 3};
+}

--- a/src/test/scala/play/api/cache/redis/connector/SerializerSpecs.scala
+++ b/src/test/scala/play/api/cache/redis/connector/SerializerSpecs.scala
@@ -5,6 +5,7 @@ import java.util.Date
 import scala.reflect.ClassTag
 
 import play.api.inject.guice.GuiceApplicationBuilder
+import play.api.cache.redis._
 
 import org.joda.time.{DateTime, DateTimeZone}
 import org.specs2.mock.Mockito
@@ -24,6 +25,11 @@ class SerializerSpecs extends Specification with Mockito {
 
     "byte" in {
       0xAB.toByte.encoded mustEqual "-85"
+      JavaTypes.byteValue.encoded mustEqual "5"
+    }
+
+    "byte[]" in {
+      JavaTypes.bytesValue.encoded mustEqual "AQID"
     }
 
     "char" in {
@@ -93,13 +99,16 @@ class SerializerSpecs extends Specification with Mockito {
           |TGlzdFNlcmlhbGl6ZUVuZCSKXGNb91MLbQIAAHhweA==
         """.stripMargin.lines.map(_.trim).mkString
     }
-
   }
 
   "AkkaDecoder" should "decode" >> {
 
     "byte" in {
       "-85".decoded[ Byte ] mustEqual 0xAB.toByte
+    }
+
+    "byte[]" in {
+      "YWJj".decoded[ Array[ Byte ] ] mustEqual Array( "a".head.toByte, "b".head.toByte, "c".head.toByte )
     }
 
     "char" in {

--- a/src/test/scala/play/api/cache/redis/impl/JavaRedisSpecs.scala
+++ b/src/test/scala/play/api/cache/redis/impl/JavaRedisSpecs.scala
@@ -1,7 +1,5 @@
 package play.api.cache.redis.impl
 
-import java.util.concurrent.Callable
-
 import scala.concurrent.duration.Duration
 
 import play.api.cache.redis._
@@ -104,6 +102,36 @@ class JavaRedisSpecs( implicit ee: ExecutionEnv ) extends Specification with Red
     "remove all" in new MockedJavaRedis {
       async.invalidate() returns execDone
       cache.removeAll().toScala must beDone.await
+    }
+
+    "get and set 'byte'" in new MockedJavaRedis {
+      val byte = JavaTypes.byteValue
+
+      // set a value
+      // note: there should be hit on "byte" but the value is wrapped instead
+      async.set( anyString, beEq( byte ), any[ Duration ] ) returns execDone
+      async.set( anyString, beEq( "byte" ), any[ Duration ] ) returns execDone
+      async.set( anyString, beEq( "java.lang.Byte" ), any[ Duration ] ) returns execDone
+      cache.set( key, byte ).toScala must beDone.await
+
+      // hit on GET
+      async.get[ Byte ]( beEq( key ) )( anyClassTag ) returns Some( byte )
+      async.get[ String ]( beEq( s"classTag::$key" ) )( anyClassTag ) returns Some( "java.lang.Byte" )
+      cache.get[ Byte ]( key ).toScala must beEqualTo( byte ).await
+    }
+
+    "get and set 'byte[]'" in new MockedJavaRedis {
+      val bytes = JavaTypes.bytesValue
+
+      // set a value
+      async.set( anyString, beEq( bytes ), any[ Duration ] ) returns execDone
+      async.set( anyString, beEq( "byte[]" ), any[ Duration ] ) returns execDone
+      cache.set( key, bytes ).toScala must beDone.await
+
+      // hit on GET
+      async.get[ Array[ Byte ] ]( beEq( key ) )( anyClassTag ) returns Some( bytes )
+      async.get[ String ]( beEq( s"classTag::$key" ) )( anyClassTag ) returns Some( "byte[]" )
+      cache.get[ Array[ Byte ] ]( key ).toScala must beEqualTo( bytes ).await
     }
   }
 }

